GGRC-2715 Migrate program roles to ACL

Smotko opened the pull request:
# TODO

- [x] Migration (might need some cleanup)
- [x] Update the frontend (Program roles order isn't ok)
- [x] Update csv files and corresponding tests
- [x] Fix failing tests (selenium tests might need a few updates)
- [x] Propagation event handlers
- [x] Audit Archiving fixes 
- [x] ACL Update to prevent Program Editors from adding ACL records

# Issue description

Migrates Programs to ACL.

Program role propagation:

```
Program
- Mapped objects (through relationships)
  - Comments & Documents
  - Proposals & Revisions
- Audit
  - Assessment Templates
  - Snapshots
  - Assessments
    - Comments & Documents
  - Issues
    - Comments & Documents
```

# Steps to test the changes

Run the migration, make sure the above propagation scheme works in all possible ways (through imports, through object creation, object deletion etc)

# Solution description

Same as Audit ACL.

# Sanity checklist

- [x] I have clicked through the app to make sure my changes work and not break the app.
- [x] I have applied the correct milestone and labels.
- [x] My changes fix the issue described in the description (and do nothing else). 🤞
- [x] My changes are covered by tests.
- [x] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [x] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [x] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).


# Migration checklist
- [ ] Migration passes all checks from our [PR review guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/reviewing_pull_requests.rst#reviewing-a-pr-containing-database-migration-scripts)


# PR Review checklist

- [ ] The changes fix the issue and don't cause any apparent regressions.
- [ ] Labels and milestone are correctly set.
- [ ] The solution description matches the changes in the code.
- [ ] There is no apparent way to improve the performance & design of the new code.
- [ ] The pull request is opened against the correct base branch.
- [ ] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

<!-- If your code is not finished yet can include a TODO check list
# TODO

- [ ] First item on the TODO
-->

AleksNeStu commented:
@Smotko I will proceed with a selenium here since the moment, when 1st ver/ of my PR w/ redesigned actual model of entities, entities_factory for RBAC will be created.

I need to finish it to provide a proper fix and avoid refactoring job.

AleksNeStu commented:
@Smotko I started work on selenium fix, ETA 4-8 hours.

AleksNeStu commented:
@Smotko please apply my quick selenium fix (contains only new UI PO changes).
@SergeyRabtsau please review it accordingly.

NOTE:
I will provide proper fix based on my PR #7125 w/ new entities factory core and w/:
- proper page objects for Program Info Widget;
- redo tests which are related to Program testing and w/ PYTHON objects instead deprecated test w/ simple collections' testing.
- new actual entity and entities factory model according new Program ACL model.

SergeyRabtsau commented:
@AleksNeStu pls update your PR [Smotko#48](https://github.com/Smotko/ggrc-core/pull/48), pls remove redundant constants that were added but not used in current implementation (PROGRAM_EDITORS, PROGRAM_READERS)

AleksNeStu commented:
@SergeyRabtsau comments were addressed

Smotko commented:
I'll merge @AleksNeStu's PR once @SergeyRabtsau gives the final thumbs up.

SergeyRabtsau commented:
@AleksNeStu, @Smotko pls be informed that selenium part looks good to me, so go ahead with applying changes from @AleksNeStu

Smotko commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,277 @@
+from ggrc.models.hooks.relationship import related
+
+
+def _get_acr_id(acl):
```
I have tried to fix this here https://github.com/google/ggrc-core/commit/861be49c86a28c0ae529bdac0eae125b34d591a4 , but because of the way Assignee and Workflow roles are propagated I couldn't include the commit here.

AlexNo commented on src/ggrc-client/js/models/simple_models.js:
```diff
@@ -93,7 +94,21 @@
+        return false;
+      }
+      const programManagerRole = GGRC.access_control_roles.find((acr) => {
+        return acr.name === 'Program Managers';
```
I'm not sure that someone won't create a role with the same name for other type of object and as a result we can get incorrect role here

Smotko replied:
Add a check for type as well.

AlexNo commented on src/ggrc-client/js/models/simple_models.js:
```diff
@@ -93,7 +94,21 @@
+      if (allowedRoles.indexOf(GGRC.current_user.system_wide_role) > -1) {
+        return false;
+      }
+      const programManagerRole = GGRC.access_control_roles.find((acr) => {
```
We already have a ticket about optimization work with ACL on FE side (GGRC-4297) and I have a [own branch](https://github.com/AlexNo/ggrc-core/tree/qf/acl) where I've tried to do something about that

Smotko commented on src/ggrc/snapshotter/acl.py:
```diff
@@ -19,7 +19,10 @@ def get_acl_payload(snapshots):
       all_models.AccessControlRole.name).filter(
       all_models.AccessControlRole.name.in_((
```
Yes, we should have a single `access_contol_role` cache, I think this will have to be done in scope of GGRC-3784 where we'll need to unify these caches anyway.

AlexNo commented on src/ggrc-client/js/mustache_helper.js:
```diff
@@ -2520,19 +2524,41 @@ Example:
   Mustache.registerHelper('is_auditor', function (options) {
     const auditor = GGRC.access_control_roles.find(
-        (role) => role.name === 'Auditors');
+      (role) => role.name === 'Auditors');
```
The same as above. Someone can create role with the same name for other object and helper can be broken

Smotko replied:
Same as above.

Smotko commented:
Hm... did we refresh revisions when we migrated workflows and audits? I don't remember us doing that so we might have the same issue as you pointed out there as well.

Smotko commented on src/ggrc/models/audit.py:
```diff
@@ -205,6 +208,19 @@ def clone(self, source_id, mapped_objects=None):
+
+    if self.archived is not None and self.archived != value and \
+       not any(acl for acl in self.access_control_list
+               if acl.ac_role.name == "Program Managers Mapped" and
```
We use strings for acr role names in a bunch of places. I'll create a ticket to move them all to constants, but I don't want to do this in this PR.

pod2metra replied:
Ok

AlexNo commented on src/ggrc-client/js/mustache_helper.js:
```diff
@@ -2520,19 +2524,41 @@ Example:
+  Mustache.registerHelper('page_roles', function (person, options) {
+    const pageInstance = Mustache.resolve(GGRC.page_instance);
+    const roles = {};
+    can.each(GGRC.access_control_roles, (role) => {
```
I'd suggest to move all the common cases to work with the `GGRC.access_control_roles` into one place so we can reuse them

Smotko replied:
Agreed. We should make it really easy to find a role by it's name, but to do it properly we'd need to refactor the way we show all custom roles and I think that's out of scope here :/

AlexNo commented on src/ggrc/assets/mustache/people/tree-item-attr.mustache:
```diff
@@ -21,48 +21,17 @@
-            <span class="role">
-              {{system_wide_role}}
+          {{log instance}}
+          {{log page_instance}}
```
please remove loggers

pod2metra commented on src/ggrc/models/hooks/acl/cache.py:
```diff
@@ -0,0 +1,28 @@
+    self.cache[key] = acl
+    if hasattr(obj, "access_control_list"):
+      obj.access_control_list.append(acl)
+    return acl
```
workflow of that method allow to add duplicates. The reason is you don't pouplate your cache for already exist objects. 
In this case i see that class as
1) singlaton (optional)
2) self.cache like dict of dicts (not default dict) 
3) add check to acl has already been created. 

```
class ACLManager(object):

  def __init__(self):
     self.object_queue = {}

  def get_or_create(self, obj, parent, person, role_id):
    if not isinstance(obj, roleadble.Roleable):
      return
    obj_key = (obj.id, obj.type)
    if obj_key not in self.object_queue:
      self.object_queue[obj_key] = {
          (acl.person_id, acl.ac_role_id, acl.parent_id): acl
          for acl in obj.access_control_list
      }
    key = (person.id, role_id, parent.id)
    if key not in self.object_queue[obj_key]:
      self.object_queue[obj_key][key] = all_models.AccessControlList(
          object=obj, parent=parent, person=person, ac_role_id=role_id)
    return self.object_queue[obj_key][key]
```

**NOTE** this snippet shows only workflow of that function and my view. It may be wrong and doesn't work correctly.

Smotko replied:
The fact that I don't prepopulate isn't a problem because I know that I'm always adding acl records that did not exist before my handler ran (handler goes over all the new acl records and uses them as the parent id on all the records it creates). I just need this 'cache' class to keep track of the records that I've created inside the handler so that I make sure that I don't create duplicates. 

I'll use the names in your suggestions, but I'll opt out to not do the prepoulation for now.

pod2metra commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,277 @@
+      acl_cache = self.caches["access_control_list_cache"]
+      acl_cache.add(obj, acl, acl.person, acl.ac_role.id)
+
+  def handle_crelationships(self, obj):
```
typo: 
handle_crelationships -> handle_relationships

pod2metra commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,277 @@
+
+def init_hook():
+  """Initialize AccessControlList-related hooks."""
+  handler = ProgramRolesHandler()
```
why do we need class here? This can be just the function, after_flush?

Smotko replied:
I need the class instance here to easily share some state between all the functions in the file.

pod2metra commented:
Do we need 
ggrc/ggrc_basic_permissions/roles/Program*.py 
ggrc/ggrc_basic_permissions/roles/Audit*.py 
files in codebase? 
It looks like that they don't grant any privileges and can be removed. Please correct me if I'm wrong.

AlexNo commented:
@Smotko value in `Authorizations` column on `People` tab of `Program` isn't changed after change the role of user ([screencast](https://drive.google.com/file/d/1PFWNYUD86XWpD0lqJT7H00EWVtSn-rcS/view?usp=sharing))

AlexNo commented:
We have the same behaviour now, but I guess that we need to report a ticket about it.
After add new person into some Program's people list on info tab, count on `People` tab isn't updated.
I'll check it with QA

AlexNo commented:
From my prospective, action when I add the person into some people list(`Program Managers`, `Program Editors` or `Program Readers`) should be the same with when I map some person on people tab and grand him the some role by `Edit Authorizations` modal.

These actions should be the equal:
**First:**
1.  Open Info tab on Program page
2. Add new person(for example `user@example.com`) into `Program Editors` list
3. Go to `People` tab
4. `user@example.com` is showed in treeview and in `Authorizations` column must be `Program Editors`
5. When I open an `Edit Authorizations` modal, `Program Editors` is selected

**Second**
1. Open `People` tab on Program page
2. Open mapper modal and map `user@example.com`
3. Open `Edit Authorizations` modal for `user@example.com` and grand him a  `Program Editors` role
4. Go to `Info` tab
5. `Program Editors` list must contain a `user@example.com`

Smotko commented on src/ggrc/views/__init__.py:
```diff
@@ -275,6 +276,19 @@ def get_access_control_roles_json():
     return as_json(published)
 
 
+def get_internal_roles_json():
```
This actually has to be refactored so that we'd only run one query instead of two.

Smotko commented:
@AlexNo We have a separate ticket to clean up the people tab. The changes in this PR are only to hold the water until then. I do need to disable/remove the authorization widget because making changes there won't really change the acl records (the only way to edit authorizations should be through the info pane).

@pod2metra yes, those will all be removed in the next rbac phases.

AlexNo commented:
Program roles are not shown on `People` tab on `Audit` page

![screen shot 2018-01-29 at 4 03 56 pm](https://user-images.githubusercontent.com/4737102/35511713-103ff374-050e-11e8-81bb-aaac2c8502b5.png)

AlexNo commented:
>  I do need to disable/remove the authorization widget because making changes there won't really change the acl records (the only way to edit authorizations should be through the info pane).
@Smotko In my opinion it may confuse users who are familiar to use the Authorizations modal window and disabling it should be within one release

Smotko commented:
@AleksNeStu between removing the people tab (what was done with workflows in the same release) or making it readonly (what I wanted to do here) I prefer having it readonly (= same as audits from 1.1 release). I'll make sure all roles are properly displayed (people that have a role an audit through assessment roles show up in the list as well, so it's not just program role propagation.)

AlexNo commented:
@Smotko looks like PR contain a migration and need to fix migration chain after merge https://github.com/google/ggrc-core/pull/7215

Smotko commented:
Ah, that was going on with all the tests, thanks for pointing it out, fixed!

AlexNo commented:
`private_program_owner` helper is no longer in use and we can remove it

pod2metra commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,288 @@
+            comment_or_document, acl, acl.person, _get_acr_id(acl))
+      return
+
+    assessment_or_issue, other = related_to(obj, {
```
Code from 218 - 231 looks the same as 233-247. Please combine them together.

Smotko replied:
If we combine them like you proposed in our hangouts meeting we will get a logical error, because for the case of Document - Issue mapping, the `mapped_to` will sometimes return the Document first and try to propagate permissions from the other (Issue), but other times it will return the issue and then propagate permissions from the other (document). If this happened then the document wouldn't get the permissions from the issue and users wouldn't be able to see the mapped documents.

I did try to refactor the code, but I ended up breaking all the propagations so it will take me a bit longer to fix it properly - I'll make a separate PR for this.

pod2metra commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,288 @@
+        acl_manager.get_or_create(assessment_or_issue, acl, acl.person,
+                                  _get_acr_id(acl))
+      return
+
```
It looks like that you can create 2 util functions here 
_create_program_mapped_roles(self, object, parent)
_create_program_roles(self, object, parent)
and use it in your handlers. 
In my opinion this will make code cleaner and reusable.

pod2metra commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,288 @@
+
+  def after_flush(self, session, _):
+    """Handle program related acl"""
+    self.access_control_list_manager = ACLManager()
```
Why you aren't doing that on __init__?

Smotko replied:
The init method is called when the flask application starts and is not rerun on each request that comes in. This means that the objects there would just keep accumulating, causing a memory leak through the life time of the application.

Having it in after_flush makes sure that the acl manager gets cleared and (hopefully) the memory gets freed up.

pod2metra commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,288 @@
+        all_models.Relationship: self.handle_relationship_creation,
+        all_models.Audit: self.handle_audit_creation
+    }
+    for obj in session.new:
```
I worry that order of instances in session.new may influence on logic. 

[Audit, Snapshot, Relationship, ACL] 
[Audit, ACL, Snapshot, Relationship]
[Audit, Relationship, ACL, Snapshot]

I think that have strict order here will be more safer:
```
objs = {type(obj): obj for obj in session.new if type(obj) in handlers}
for obj in objs.get(all_models.Audit, []):
  if obj not in self.done:
    handlers[all_models.Audit](obj)

for obj in objs.get(all_models.AccessControlList, []):
  if obj not in self.done:
    handlers[all_models.AccessControlList](obj)
...
```

Smotko replied:
As discussed in the offline, the ACLManager should handle all the cases for this.

pod2metra commented:
After migration running I have few issues

1) No info about program mapped roles. 
```
select r.name, r.object_type, l.object_type, count(1) from access_control_roles as r join access_control_list as l on l.ac_role_id = r.id where r.name like "%Program%" group by 1,2,3;
+------------------+-------------+-------------+----------+
| name             | object_type | object_type | count(1) |
+------------------+-------------+-------------+----------+
| Program Editors  | Program     | Program     |      197 |
| Program Managers | Program     | Program     |     1124 |
| Program Readers  | Program     | Program     |      230 |
+------------------+-------------+-------------+----------+
``` 

2) All roles related only to Program object type. 
I think that should be created separate roles:
Program Managers Mapped to mapped objects (Snapshots, Comments, ... ) and Program roles to program only. 
Please correct my if I'm wrong.
```
mysql> select id, name, object_type from access_control_roles where name like "%Program%";
+-----+-------------------------+-------------+
| id  | name                    | object_type |
+-----+-------------------------+-------------+
| 103 | Program Editors         | Program     |
| 106 | Program Editors Mapped  | Program     |
| 102 | Program Managers        | Program     |
| 105 | Program Managers Mapped | Program     |
| 104 | Program Readers         | Program     |
| 107 | Program Readers Mapped  | Program     |
+-----+-------------------------+-------------+
6 rows in set (0.00 sec)
```

pod2metra commented:
**Nice to have.** Will be good if we will have a separate handle/api/background task that will recalculate ACL if we have some inconsistency in DB. 
This can be done in separate PR.

AlexNo commented:
@Smotko I've also noticed that any of Audits for example don't contain people with Program's roles, after run migration.

Smotko commented:
I have fixed the program roles propagation migration, can you check it again? Thanks!

Smotko commented:
@pod2metra about point 2 of your migration comment, we tend to use the parent object types for propagated mapped roles except if the mapped role applies to only one object type. I believe this is because how assessment propagation was implemented and then audits, workflows and now programs used the same pattern.

pod2metra commented:
@Smotko it looks confusing. The role for object_type Program assign to Control for example. 
Also we lose easy possibility to change permissions to object. 
For example for program editor we should change permissions for control only from RU to R.
In this case it required some changes in python code and complex datatmigration. 
If we have separate roles to each model this will be only one simple datamigration (easy in develop, easy to check, easy to review, easy to deploy).
**This isn't the blocker, but I think that have separate roles to each model will be much better.**

Smotko commented:
@pod2metra I agree completely with you, but this is the sort of a PR comment we should have had on the Assessment propagation PR or at the audit propagation PR or at the workflow propagation PR. Tackling this now without fixing the other cases where we made this mistake doesn't really make sense 😞

Smotko commented on src/ggrc/models/hooks/acl/program_roles.py:
```diff
@@ -0,0 +1,288 @@
+    role_map = self.program_roles
+    program, other = related_to(obj, "Program")
+    if program:
+      for acl in program.access_control_list:
```
Same as above.

pod2metra commented:
@Smotko i found exception in your migration. 
Share db dump with you. 

sqlalchemy.exc.IntegrityError: (IntegrityError) (1062, "Duplicate entry '4-107-9-Control-337772' for key 'person_id'") "\n    INSERT INTO access_control_list(\n        person_id, ac_role_id, object_id, object_type,\n        created_at, updated_at, context_id, parent_id)\n    SELECT acl.person_id, acrn.id, r.source_id, r.source_type,\n           acl.created_at, acl.updated_at, acl.context_id, acl.id\n    FROM access_control_list AS acl\n    JOIN access_control_roles AS acr ON acr.id = acl.ac_role_id\n     AND acr.name = 'Program Readers'\n    JOIN access_control_roles as acrn ON acrn.name = 'Program Readers Mapped'\n    JOIN relationships AS r\n      ON r.destination_id = acl.object_id\n     AND r.destination_type = acl.object_type;\n    " ()

AlexNo commented:
@Smotko don't forget to remove debug commit https://github.com/google/ggrc-core/pull/7076/commits/ff39b15f21910ca27f049976b2cc237a9550bb90

AlexNo commented:
Looks like `Authorization` field on info pane of Person is broken for ACL roles.
For calculation  of role is using `with_program_roles_as` helper which uses `UserRole`.

![screen shot 2018-01-30 at 5 44 27 pm](https://user-images.githubusercontent.com/4737102/35572243-4786a10c-05e5-11e8-9aaf-2fcb9b95ecb9.png)


I guess helper is already outdated, and we need to remove it or change the logic there

AlexNo commented on src/ggrc/assets/mustache/people/tree_add_item.mustache:
```diff
@@ -2,21 +2,3 @@
     Copyright (C) 2018 Google Inc.
```
I guess that we can remove this file at all

pod2metra commented:
Looks like I find the issue: 
1) create 3 users (program_manager, program_editor, program_reader)
2) create program as program_manager 
3) add acl to editor and reader 
4) create control as program_manager and map it to program 
5) login as program editor 
6) create control and map it to program 
7) login as program_manager 

Open changelog: 

  File "/vagrant-dev/opt/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 267, in Handle
    result = handler(dict(self._environ), self._StartResponse)
  File "/vagrant/src/packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/vagrant/src/packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/vagrant/src/packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/vagrant/src/packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/vagrant/src/packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/vagrant/src/packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/vagrant/src/packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/vagrant/src/packages/flask_login.py", line 758, in decorated_view
    return func(*args, **kwargs)
  File "/vagrant/src/packages/flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 1484, in dispatch_request
    return super(ReadOnlyResource, self).dispatch_request(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 608, in dispatch_request
    return self.collection_get()
  File "/vagrant/src/ggrc/services/common.py", line 956, in collection_get
    objs = filter_resource(objs)
  File "/vagrant/src/ggrc/services/common.py", line 1542, in filter_resource
    sub_resource, depth=depth + 1, user_permissions=user_permissions)
  File "/vagrant/src/ggrc/services/common.py", line 1606, in filter_resource
    value, depth=depth + 1, user_permissions=user_permissions)
  File "/vagrant/src/ggrc/services/common.py", line 1567, in filter_resource
    inst = resource[name]
KeyError: 'source'

pod2metra commented:
Question: 
Should program manager see on `http://127.0.0.1:8080/dashboard#!control_widget` page control, created by another user, but mapped to his program by another user?

zidarsk8 commented:
@pod2metra nope, on my work page the user should only see items that he has a direct role on (a user visible role, not mapped roles) But the documentation on taht is now a bit old since it also includes "no role" mapping to a program that we do not support anymore.

pod2metra commented:
Issue on opening tab `define assessment template`. 

<img width="279" alt="screen shot 2018-01-30 at 23 39 37" src="https://user-images.githubusercontent.com/2368067/35590121-f1f9c624-0616-11e8-9883-d3f932f9c7fc.png">
<img width="639" alt="screen shot 2018-01-30 at 23 39 12" src="https://user-images.githubusercontent.com/2368067/35590126-f43df932-0616-11e8-8573-f2f2ed31d22b.png">

FE error, but it makes unable to create Assessment Template over FE.

Smotko commented:
@pod2metra I'll look into the first issue, I'm not 100% if it's related to my changes, but I'll check.

I'll also take a look at the second one, but the error stack seems to be identical to: GGRC-4179.

Smotko commented:
I have confirmed that the second issue is the same as GGRC-4179 and is reproducible on ggrc-test as well. I have added a note to the ticket that it happens everytime assessment edit/new modal is opened not on audit page.

zidarsk8 commented:
@Smotko I am not sure why but on this PR I get a migration warning

```# db_migrate 
INFO     2018-01-30 21:17:45,241 ggrc.migrate Upgrading ggrc
INFO  [alembic.migration] Context impl MySQLImpl.
INFO  [alembic.migration] Will assume non-transactional DDL.
INFO  [alembic.migration] Context impl MySQLImpl.
INFO  [alembic.migration] Will assume non-transactional DDL.
INFO  [alembic.migration] Running upgrade 568f400e4a62 -> 242b8dc8493b, Migrate program roles to ACL
INFO  [alembic.migration] Context impl MySQLImpl.
INFO  [alembic.migration] Will assume non-transactional DDL.
INFO  [alembic.migration] Context impl MySQLImpl.
INFO  [alembic.migration] Will assume non-transactional DDL.
INFO  [alembic.migration] Context impl MySQLImpl.
INFO  [alembic.migration] Will assume non-transactional DDL.
INFO  [alembic.migration] Context impl MySQLImpl.
INFO  [alembic.migration] Will assume non-transactional DDL.
/vagrant/src/packages/sqlalchemy/sql/selectable.py:1214: SAWarning: Column 'permissions_json' on table <sqlalchemy.sql.selectable.TableClause at 0x7ff1542d19d0; roles> being replaced by <sqlalchemy.sql.elements.ColumnClause at 0x7ff1542d17d0; permissions_json>, which has the same key.  Consider use_labels for select() statements.
  self._columns[c.key] = c

```

Smotko commented:
@pod2metra first issue is reproducible on the dev branch so we'll tackle it in a separate PR as well.

Smotko commented:
@zidarsk8 good find, the issue was in my update to the workflow migration where the permission_json field was defined twice:

```
+roles = sa.sql.table(
 +    'roles',
 +    sa.sql.column('name', sa.String),
 +    sa.sql.column('permissions_json', sa.String),
 +    sa.sql.column('description', sa.String),
 +    sa.sql.column('permissions_json', sa.String),
 +    sa.sql.column('created_at', sa.DateTime),
 +    sa.sql.column('updated_at', sa.DateTime),
 +    sa.sql.column('scope', sa.String),
 +)
```
I have updated the commit and the warning should now be gone.

AlexNo commented:
After edit a Program in an
![screen shot 2018-01-31 at 10 45 22 am](https://user-images.githubusercontent.com/4737102/35610932-47bc4c8c-0674-11e8-87a5-430793afc28e.png)
 Edit modal, `Submit for review` button is displayed twice

AlexNo commented:
We have different order of role's controls in `Create Program` modal and `Edit Program` modal with info pane of `Program`

**Create Program**
![screen shot 2018-01-31 at 10 57 16 am](https://user-images.githubusercontent.com/4737102/35611403-e805cca8-0675-11e8-8282-9bced840ef90.png)

**Edit Program**
![screen shot 2018-01-31 at 10 58 06 am](https://user-images.githubusercontent.com/4737102/35611401-e7c8bff2-0675-11e8-8994-a5289769a9d1.png)
**Info pane**
![screen shot 2018-01-31 at 10 57 53 am](https://user-images.githubusercontent.com/4737102/35611402-e7e54730-0675-11e8-8f2d-116592ce6d2d.png)

I guess that it isn't good from UX perspective

Smotko commented:
@AlexNo I have fixed the order in new program modal, but I can't seem to reproduce the submit for review issue. Are there any additional steps that I'm missing? Here is a video of me trying to reproduce locally:

![submitforreview](https://user-images.githubusercontent.com/513444/35612615-aa83b93c-0669-11e8-9d0f-2dbabecce9f2.gif)

AlexNo commented:
@Smotko It's reproduced constantly when I'm editing a CA, but seems issue is not related with your changes, because it also is reproduced on `dev` instance.
I'll check it with QA

pod2metra commented:
Program manager doesn't see `open audit` link on http://127.0.0.1:8080/programs/1#!audit_widget page. 
But he can open it http://127.0.0.1:8080/audits/1#!Summary_widget and see the content of the audit.

pod2metra commented:
Program manager can't generate assessment over assessment generation procedure to the audit there he has `Program Managers Mapped` roles. 
WARNING  2018-01-31 08:55:19,289 common.py:1283] Collection POST commit failed
Traceback (most recent call last):
  File "/vagrant/src/ggrc/services/common.py", line 1277, in collection_post
    self.collection_post_loop(body, res, no_result)
  File "/vagrant/src/ggrc/services/common.py", line 1186, in collection_post_loop
    self._check_post_permissions(objects)
  File "/vagrant/src/ggrc/services/common.py", line 1133, in _check_post_permissions
    raise Forbidden()
Forbidden: 403 Forbidden: You don't have the permission to access the requested resource. It is either read-protected or not readable by the server.

Smotko commented:
@pod2metra nice finds. I believe we had an open ticket for the missing object links, but since the fix isn't that difficult I have included it here. I have also fixed the assessment generation for creators.

AlexNo commented:
@Smotko looks like `Change log` tab is broken for Program, could you take a look?
```ERROR    2018-01-31 12:38:03,858 ggrc.services.common 'list' object has no attribute 'iteritems'
Traceback (most recent call last):
  File "/vagrant/src/ggrc/services/common.py", line 608, in dispatch_request
    return self.collection_get()
  File "/vagrant/src/ggrc/services/common.py", line 977, in collection_get
    collection, self.collection_last_modified(), cache_op=cache_op)
  File "/vagrant/src/ggrc/services/common.py", line 1455, in json_success_response
    (self.as_json(response_object), status, headers))
  File "/vagrant/src/ggrc/services/common.py", line 1343, in as_json
    return as_json(obj, **kwargs)
  File "/vagrant/src/ggrc/utils/__init__.py", line 53, in as_json
    return json.dumps(obj, cls=GrcEncoder, **kwargs)
  File "/usr/lib/python2.7/json/__init__.py", line 251, in dumps
    sort_keys=sort_keys, **kw).encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "/vagrant/src/ggrc/utils/__init__.py", line 47, in default
    return obj()
  File "/vagrant/src/ggrc/models/revision.py", line 101, in foo
    return revisions_diff.prepare(instance, self.content)
  File "/vagrant/src/ggrc/utils/revisions_diff/builder.py", line 264, in prepare
    diff_data.pop("custom_attributes", None),
  File "/vagrant/src/ggrc/utils/revisions_diff/builder.py", line 148, in generate_cav_diff
    instance.custom_attribute_definitions)
  File "/vagrant/src/ggrc/utils/revisions_diff/builder.py", line 120, in populate_cavs
    custom_attributes = {int(k): v for k, v in custom_attributes.iteritems()}
AttributeError: 'list' object has no attribute 'iteritems'
INFO     2018-01-31 12:38:04,000 werkzeug 172.19.0.1 - - [31/Jan/2018 12:38:04] "GET /api/revisions?__sort=updated_at&resource_type=Program&resource_id=4&_=1517402276793 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/vagrant-dev/opt/gae_packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/vagrant-dev/opt/gae_packages/flask_login.py", line 758, in decorated_view
    return func(*args, **kwargs)
  File "/vagrant-dev/opt/gae_packages/flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 1484, in dispatch_request
    return super(ReadOnlyResource, self).dispatch_request(*args, **kwargs)
  File "/vagrant/src/ggrc/services/common.py", line 608, in dispatch_request
    return self.collection_get()
  File "/vagrant/src/ggrc/services/common.py", line 977, in collection_get
    collection, self.collection_last_modified(), cache_op=cache_op)
  File "/vagrant/src/ggrc/services/common.py", line 1455, in json_success_response
    (self.as_json(response_object), status, headers))
  File "/vagrant/src/ggrc/services/common.py", line 1343, in as_json
    return as_json(obj, **kwargs)
  File "/vagrant/src/ggrc/utils/__init__.py", line 53, in as_json
    return json.dumps(obj, cls=GrcEncoder, **kwargs)
  File "/usr/lib/python2.7/json/__init__.py", line 251, in dumps
    sort_keys=sort_keys, **kw).encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "/vagrant/src/ggrc/utils/__init__.py", line 47, in default
    return obj()
  File "/vagrant/src/ggrc/models/revision.py", line 101, in foo
    return revisions_diff.prepare(instance, self.content)
  File "/vagrant/src/ggrc/utils/revisions_diff/builder.py", line 264, in prepare
    diff_data.pop("custom_attributes", None),
  File "/vagrant/src/ggrc/utils/revisions_diff/builder.py", line 148, in generate_cav_diff
    instance.custom_attribute_definitions)
  File "/vagrant/src/ggrc/utils/revisions_diff/builder.py", line 120, in populate_cavs
    custom_attributes = {int(k): v for k, v in custom_attributes.iteritems()}
AttributeError: 'list' object has no attribute 'iteritems'
```

pod2metra commented:
@AlexNo @Smotko that issue fixed in this pr https://github.com/google/ggrc-core/pull/7204

pod2metra commented:
Program manager doesn't see create audit button on http://127.0.0.1:8080/audits/1#!assessment_widget 
He can only `generate assessment`.

program manager page presentation
![screen shot 2018-01-31 at 5 42 01 pm](https://user-images.githubusercontent.com/2368067/35628846-27d0321e-06ae-11e8-9746-5715e4889f5d.png)

admin page presentation 
![screen shot 2018-01-31 at 5 42 11 pm](https://user-images.githubusercontent.com/2368067/35628867-3455c2e2-06ae-11e8-861c-7f0a52d16184.png)

pod2metra commented:
**Question**
After assessment generation created duplicate ACL for assessment
```
mysql> select l.id, r.name, l.person_id, l.object_type, l.object_id, l.parent_id from access_control_list as l join access_control_roles as r on r.id = l.ac_role_id where r.name like 'Program % Mapped' and l.object_type = "Assessment";
+----+-------------------------+-----------+-------------+-----------+-----------+
| id | name                    | person_id | object_type | object_id | parent_id |
+----+-------------------------+-----------+-------------+-----------+-----------+
| 43 | Program Editors Mapped  |         3 | Assessment  |         1 |        26 |
| 46 | Program Editors Mapped  |         3 | Assessment  |         1 |        18 |
| 42 | Program Managers Mapped |         2 | Assessment  |         1 |        23 |
| 45 | Program Managers Mapped |         2 | Assessment  |         1 |        17 |
| 44 | Program Readers Mapped  |         4 | Assessment  |         1 |        29 |
| 47 | Program Readers Mapped  |         4 | Assessment  |         1 |        19 |
+----+-------------------------+-----------+-------------+-----------+-----------+
```

here is their parents 
```
mysql> select l.id, r.name, l.person_id, l.object_type, l.object_id, l.parent_id from access_control_list as l join access_control_roles as r on r.id = l.ac_role_id where
    -> l.id in (select l.parent_id from access_control_list as l join access_control_roles as r on r.id = l.ac_role_id where r.name like 'Program % Mapped' and l.object_type = "Assessment");
+----+-------------------------+-----------+-------------+-----------+-----------+
| id | name                    | person_id | object_type | object_id | parent_id |
+----+-------------------------+-----------+-------------+-----------+-----------+
| 26 | Program Editors Mapped  |         3 | Snapshot    |         1 |        18 |
| 18 | Program Editors Mapped  |         3 | Audit       |         1 |         2 |
| 23 | Program Managers Mapped |         2 | Snapshot    |         1 |        17 |
| 17 | Program Managers Mapped |         2 | Audit       |         1 |         1 |
| 29 | Program Readers Mapped  |         4 | Snapshot    |         1 |        19 |
| 19 | Program Readers Mapped  |         4 | Audit       |         1 |         3 |
+----+-------------------------+-----------+-------------+-----------+-----------+
```
It looks strange, i think that permissions to assessment should be grant by Audit but not from snapshots, correct me please here if I am wrong.

pod2metra commented:
When you create label as program editor/manager after update you will not see that label. 
ACL didn't grant to the labels.
Program role page presentation 
![screen shot 2018-01-31 at 6 11 59 pm](https://user-images.githubusercontent.com/2368067/35630373-506596d4-06b2-11e8-9e4d-62fd67f65172.png)
Admin role page presentation
![screen shot 2018-01-31 at 6 11 34 pm](https://user-images.githubusercontent.com/2368067/35630394-590a96fe-06b2-11e8-9000-9f8eeb6a91e1.png)

pod2metra commented:
After mark one assessment reuse another one will be repopulated ACLs between assessments (ACLs for documents for example will be duplicated).

Smotko commented:
I have addressed all the comments above except the labels one, because I can't reproduce it locally.  I have looked into it and we never create acl records for labels so the permissions aren't propagated in a way that this pull request could affect them (if there is an issue other roles like Assessment Assignees, Auditors, ...) will also have the same issue and we can tackle it separately.

NastassiaKiseleva commented:
@Smotko @pod2metra Please create a ticket to handle labels issue separately, and hopefully we are good to go!

pod2metra commented:
On the same snapshot ACL permissions grant twice: 

One from Audit second from Assessment 
| 17 | Program Managers Mapped |         3 | Audit       |         1 |         1 |
| 18 | Program Editors Mapped  |         2 | Audit       |         1 |         2 |
| 19 | Program Readers Mapped  |         5 | Audit       |         1 |         3 |
| 23 | Program Managers Mapped |         3 | Snapshot    |         1 |        17 |
| 24 | Program Managers Mapped |         3 | Snapshot    |         2 |        17 |
| 25 | Program Managers Mapped |         3 | Snapshot    |         3 |        17 |
| 26 | Program Editors Mapped  |         2 | Snapshot    |         1 |        18 |
| 27 | Program Editors Mapped  |         2 | Snapshot    |         2 |        18 |
| 28 | Program Editors Mapped  |         2 | Snapshot    |         3 |        18 |
| 29 | Program Readers Mapped  |         5 | Snapshot    |         1 |        19 |
| 30 | Program Readers Mapped  |         5 | Snapshot    |         2 |        19 |
| 31 | Program Readers Mapped  |         5 | Snapshot    |         3 |        19 |
| 62 | Program Managers Mapped |         3 | Assessment  |         3 |        17 |
| 63 | Program Editors Mapped  |         2 | Assessment  |         3 |        18 |
| 64 | Program Readers Mapped  |         5 | Assessment  |         3 |        19 |
| 65 | Program Managers Mapped |         3 | Assessment  |         1 |        17 |
| 66 | Program Editors Mapped  |         2 | Assessment  |         1 |        18 |
| 67 | Program Readers Mapped  |         5 | Assessment  |         1 |        19 |
| 68 | Program Managers Mapped |         3 | Assessment  |         2 |        17 |
| 69 | Program Editors Mapped  |         2 | Assessment  |         2 |        18 |
| 70 | Program Readers Mapped  |         5 | Assessment  |         2 |        19 |
| 73 | Program Managers Mapped |         3 | Snapshot    |         3 |        62 |
| 74 | Program Editors Mapped  |         2 | Snapshot    |         3 |        63 |
| 75 | Program Readers Mapped  |         5 | Snapshot    |         3 |        64 |
| 76 | Program Managers Mapped |         3 | Snapshot    |         2 |        65 |
| 77 | Program Editors Mapped  |         2 | Snapshot    |         2 |        66 |
| 78 | Program Readers Mapped  |         5 | Snapshot    |         2 |        67 |
+----+-------------------------+-----------+-------------+-----------+-----------+

Moreover duplicate ACL grant only for 2 and 3 snapshot but for 1. 
Procedure was autogenerate Assessment from Audit mapped controls.

pod2metra commented:
Label issue fixed, but it looks like side effect of another fix. Will be good if it is covered by test. 
(Note the issue of that problem may be memcache) 
PS: this can be done in separate PR.

pod2metra commented:
**That issue is still valid:** 
Program manager doesn't see create audit button on http://127.0.0.1:8080/audits/1#!assessment_widget 

He can only `generate assessment`.

program manager page presentation
![screen shot 2018-01-31 at 5 42 01 pm](https://user-images.githubusercontent.com/2368067/35628846-27d0321e-06ae-11e8-9746-5715e4889f5d.png)

admin page presentation 
![screen shot 2018-01-31 at 5 42 11 pm](https://user-images.githubusercontent.com/2368067/35628867-3455c2e2-06ae-11e8-861c-7f0a52d16184.png)

pod2metra commented:
1) Over import I can't setup empty program editor/reader list (please add ability to setup empty list and add how to do that in description).
This can be done separately.
